staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go

```diff
@@ -0,0 +1,257 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"math/rand"
+	"testing"
+)
+
+func TestValidateParameters(t *testing.T) {
+	tests := []struct {
+		name      string
+		deckSize  int32
+		handSize  int32
+		validated bool
+	}{
+		{
+			"deckSize is < 0",
+			-100,
+			8,
+			false,
+		},
+		{
+			"handSize is < 0",
+			128,
+			-100,
+			false,
+		},
+		{
+			"deckSize is 0",
+			0,
+			8,
+			false,
+		},
+		{
+			"handSize is 0",
+			128,
+			0,
+			false,
+		},
+		{
+			"handSize is greater than deckSize",
+			128,
+			129,
+			false,
+		},
+		{
+			"deckSize: 128 handSize: 6",
+			128,
+			6,
+			true,
+		},
+		{
+			"deckSize: 1024 handSize: 6",
+			1024,
+			6,
+			true,
+		},
+		{
+			"deckSize: 512 handSize: 8",
+			512,
+			8,
+			false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if ValidateParameters(test.deckSize, test.handSize) != test.validated {
+				t.Errorf("test case %s fails", test.name)
+				return
+			}
+		})
+	}
+}
+
+func BenchmarkValidateParameters(b *testing.B) {
+	deckSize, handSize := int32(512), int32(8)
+	for i := 0; i < b.N; i++ {
+		_ = ValidateParameters(deckSize, handSize)
+	}
+}
+
+func TestShuffleAndDealWithValidation(t *testing.T) {
+	tests := []struct {
+		name      string
+		deckSize  int32
+		handSize  int32
+		pick      func(int32)
+		validated bool
+	}{
+		{
+			"deckSize is < 0",
+			-100,
+			8,
+			func(int32) {},
+			false,
+		},
+		{
+			"handSize is < 0",
+			128,
+			-100,
+			func(int32) {},
+			false,
+		},
+		{
+			"deckSize is 0",
+			0,
+			8,
+			func(int32) {},
+			false,
+		},
+		{
+			"handSize is 0",
+			128,
+			0,
+			func(int32) {},
+			false,
+		},
+		{
+			"handSize is greater than deckSize",
+			128,
+			129,
+			func(int32) {},
+			false,
+		},
+		{
+			"deckSize: 128 handSize: 6",
+			128,
+			6,
+			func(int32) {},
+			true,
+		},
+		{
+			"deckSize: 1024 handSize: 6",
+			1024,
+			6,
+			func(int32) {},
+			true,
+		},
+		{
+			"deckSize: 512 handSize: 8",
+			512,
+			8,
+			func(int32) {},
+			false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if (ShuffleAndDealWithValidation(rand.Uint64(), test.deckSize, test.handSize, test.pick) == nil) != test.validated {
+				t.Errorf("test case %s fails", test.name)
+				return
+			}
+		})
+	}
+}
+
+func BenchmarkShuffleAndDeal(b *testing.B) {
+	hashValue := rand.Uint64()
+	deckSize, handSize := int32(512), int32(8)
+	pick := func(int32) {}
+	for i := 0; i < b.N; i++ {
+		ShuffleAndDeal(hashValue, deckSize, handSize, pick)
```

Comment:
How about verying the hashValue?  For example, using `i` as the hashValue?

Reply:
Good suggestion.